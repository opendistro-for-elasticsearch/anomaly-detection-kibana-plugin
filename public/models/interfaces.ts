/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { DATA_TYPES, DETECTOR_STATE } from '../utils/constants';

export type FieldInfo = {
  label: string;
  type: DATA_TYPES;
};

export enum FILTER_TYPES {
  SIMPLE = 'simple_filter',
  CUSTOM = 'custom_filter',
}

export enum FEATURE_TYPE {
  SIMPLE = 'simple_aggs',
  CUSTOM = 'custom_aggs',
}

export enum OPERATORS_MAP {
  IS = 'is',
  IS_NOT = 'is_not',
  IS_NULL = 'is_null',
  IS_NOT_NULL = 'is_not_null',
  IS_GREATER = 'is_greater',
  IS_GREATER_EQUAL = 'is_greater_equal',
  IS_LESS = 'is_less',
  IS_LESS_EQUAL = 'is_less_equal',
  STARTS_WITH = 'starts_with',
  ENDS_WITH = 'ends_with',
  CONTAINS = 'contains',
  NOT_CONTAINS = 'does_not_contains',
  IN_RANGE = 'in_range',
  NOT_IN_RANGE = 'not_in_range',
}

export type UIFilter = {
  fieldInfo: FieldInfo[];
  operator: OPERATORS_MAP;
  fieldValue?: string | number | null;
  fieldRangeStart?: number;
  fieldRangeEnd?: number;
};

export type FeatureAttributes = {
  featureId?: string;
  featureName: string;
  featureEnabled: boolean;
  importance: number;
  aggregationQuery: { [key: string]: any };
};

export enum UNITS {
  MINUTES = 'MINUTES',
}

export type Schedule = {
  interval: number;
  unit: UNITS;
};

export type UiFeature = {
  featureType: FEATURE_TYPE;
  aggregationBy?: string;
  aggregationOf?: string;
};

export type UiMetaData = {
  filterType: FILTER_TYPES;
  filters: UIFilter[];
  features: {
    [key: string]: UiFeature;
  };
};
export type Detector = {
  primaryTerm: number;
  seqNo: number;
  id: string;
  name: string;
  description: string;
  timeField: string;
  indices: string[];
  filterQuery: { [key: string]: any };
  featureAttributes: FeatureAttributes[];
  windowDelay: { period: Schedule };
  detectionInterval: { period: Schedule };
  uiMetadata: UiMetaData;
  lastUpdateTime: number;
  enabled?: boolean;
  enabledTime?: number;
  disabledTime?: number;
  curState: DETECTOR_STATE;
  initializationError: string;
};

export type DetectorListItem = {
  id: string;
  name: string;
  indices: string[];
  curState: DETECTOR_STATE;
  featureAttributes: FeatureAttributes[];
  totalAnomalies: number;
  lastActiveAnomaly: number;
  lastUpdateTime: number;
};

export type AnomalyData = {
  anomalyGrade: number;
  anomalyScore?: number;
  confidence: number;
  detectorId?: string;
  endTime: number;
  startTime: number;
  plotTime?: number;
};

export type FeatureAggregationData = {
  data: number;
  endTime: number;
  startTime: number;
  plotTime?: number;
};

export type Anomalies = {
  anomalies: AnomalyData[];
  featureData: { [key: string]: FeatureAggregationData[] };
};

export type Monitor = {
  id: string;
  name: string;
  enabled: boolean;
  enabledTime?: number;
  schedule: { period: Schedule };
  inputs: any[];
  triggers: any[];
  lastUpdateTime: number;
};

export type MonitorAlert = {
  monitorName: string;
  triggerName: string;
  severity: number;
  state: string;
  error: string;
  startTime: number;
  endTime: number;
  acknowledgedTime: number;
};

export type AnomalySummary = {
  anomalyOccurrence: number;
  minAnomalyGrade: number;
  maxAnomalyGrade: number;
  minConfidence: number;
  maxConfidence: number;
  lastAnomalyOccurrence: string;
};

export type DateRange = {
  startDate: number;
  endDate: number;
};
