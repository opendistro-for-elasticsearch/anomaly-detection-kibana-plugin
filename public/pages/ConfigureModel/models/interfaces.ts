/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}
