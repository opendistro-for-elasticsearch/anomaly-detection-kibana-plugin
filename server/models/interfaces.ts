/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export interface DefaultHeaders {
  'Content-Type': 'application/json';
  Accept: 'application/json';
  'User-Agent': 'Kibana';
}

export interface SearchResponse<T> {
  hits: {
    total: { value: number };
    hits: Array<{
      _source: T;
      _id: string;
      _seq_no?: number;
      _primary_term?: number;
    }>;
  };
}

export interface ADApis {
  [API_ROUTE: string]: string;
  readonly DETECTOR_BASE: string;
}

export interface AlertingApis {
  [API_ROUTE: string]: string;
  readonly ALERTING_BASE: string;
}
export interface Entity {
  name: string;
  value: string;
}
export interface Anomaly {
  anomalyGrade: number;
  confidence: number;
  anomalyScore: number;
  startTime: number;
  endTime: number;
  plotTime: number;
  entity?: Entity[];
}
// Plot time is middle of start and end time to provide better visualization to customers
// Example, if window is 10 mins, in a given startTime and endTime of 12:10 to 12:20 respectively.
// plotTime will be 12:15.
export interface FeatureData {
  startTime: number;
  endTime: number;
  plotTime: number;
  data: number;
}
export interface AnomalyResults {
  anomalies: Anomaly[];
  featureData: { [key: string]: FeatureData[] };
}

export interface InitProgress {
  percentageStr: string;
  estimatedMinutesLeft: number;
  neededShingles: number;
}

export interface EntityAnomalySummary {
  startTime: number;
  maxAnomaly: number;
  anomalyCount: number;
}

export interface EntityAnomalySummaries {
  entity: Entity;
  anomalySummaries: EntityAnomalySummary[];
}
