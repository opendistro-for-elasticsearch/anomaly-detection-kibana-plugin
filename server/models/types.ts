/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { SORT_DIRECTION } from '../utils/constants';

export type CatIndex = {
  index: string;
  health: string;
};

export type IndexAlias = {
  index: string;
  alias: string;
};

export type GetAliasesResponse = {
  aliases: IndexAlias[];
};

export type GetIndicesResponse = {
  indices: CatIndex[];
};

export type GetMappingResponse = {
  [key: string]: any;
};

export enum UNITS {
  MINUTES = 'MINUTES',
}

export type Schedule = {
  interval: number;
  unit: UNITS;
};

export type FeatureAttributes = {
  featureId?: string;
  featureName: string;
  featureEnabled: boolean;
};

export type Detector = {
  id?: string;
  name: string;
  description: string;
  indices: string[];
  filterQuery?: { [key: string]: any };
  featureAttributes?: FeatureAttributes[];
  windowDelay?: { period: Schedule };
  detectionInterval?: { period: Schedule };
  uiMetadata?: { [key: string]: any };
  lastUpdateTime: number;
  enabled: boolean;
  enabledTime?: number;
  disabledTime?: number;
};

export type Monitor = {
  id: string;
  name: string;
  enabled: boolean;
  enabledTime?: number;
  schedule: { period: Schedule };
  inputs: any[];
  triggers: any[];
  lastUpdateTime: number;
};

export type GetDetectorsQueryParams = {
  from: number;
  size: number;
  search: string;
  indices?: string;
  sortDirection: SORT_DIRECTION;
  sortField: string;
};

export type GetAdMonitorsQueryParams = {
  from: number;
  size: number;
  search: string;
  indices?: string;
  sortDirection: SORT_DIRECTION;
  sortField: string;
};

export type DetectorResultsQueryParams = {
  from: number;
  size: number;
  sortDirection: SORT_DIRECTION;
  sortField: string;
  dateRangeFilter?: DateRangeFilter;
};

export type AnomalyResult = {
  startTime: number;
  endTime: number;
  plotTime: number;
  anomalyGrade: number;
  // grade: number;
  confidence: number;
};

export type FeatureResult = {
  startTime: number;
  endTime: number;
  plotTime: number;
  data: number;
};

export type AnomalyResultsResponse = {
  totalAnomalies: number;
  results: AnomalyResult[];
  featureResults: { [key: string]: FeatureResult[] };
};

export type ServerResponse<T> =
  | { ok: false; error: string }
  | { ok: true; response: T };

export type DateRangeFilter = {
  startTime?: number;
  endTime?: number;
  fieldName: string;
}