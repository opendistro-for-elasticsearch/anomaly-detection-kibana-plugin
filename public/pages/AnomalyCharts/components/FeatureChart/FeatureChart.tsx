/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  Chart,
  Axis,
  LineSeries,
  RectAnnotation,
  niceTimeFormatter,
  Position,
  Settings,
  ScaleType,
} from '@elastic/charts';
import { EuiEmptyPrompt, EuiText, EuiLink, EuiButton } from '@elastic/eui';
import React, { useState, Fragment } from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { useDelayedLoader } from '../../../../hooks/useDelayedLoader';
import {
  FeatureAggregationData,
  FeatureAttributes,
  DateRange,
  FEATURE_TYPE,
} from '../../../../models/interfaces';
import { darkModeEnabled } from '../../../../utils/kibanaUtils';
import { prepareDataForChart } from '../../../utils/anomalyResultUtils';
import { CodeModal } from '../../../DetectorConfig/components/CodeModal/CodeModal';
import { CHART_FIELDS, CHART_COLORS, FEATURE_CHART_THEME } from '../../utils/constants';

interface FeatureChartProps {
  feature: FeatureAttributes;
  featureData: FeatureAggregationData[];
  annotations: any[];
  isLoading: boolean;
  dateRange: DateRange;
  featureType: FEATURE_TYPE;
  field?: string;
  aggregationMethod?: string;
  aggregationQuery?: string;
  featureDataSeriesName: string;
  edit?: boolean;
  onEdit?(): void;
}
const getDisabledChartBackground = () =>
  darkModeEnabled() ? '#25262E' : '#F0F0F0';

export const FeatureChart = (props: FeatureChartProps) => {
  const [showCustomExpression, setShowCustomExpression] = useState<boolean>(
    false
  );
  const timeFormatter = niceTimeFormatter([
    props.dateRange.startDate,
    props.dateRange.endDate,
  ]);
  const showLoader = useDelayedLoader(props.isLoading);

  const featureDescription = () => (
    <EuiText size="s">
      {props.featureType === FEATURE_TYPE.SIMPLE ? (
        <Fragment>
          <span
            className="content-panel-subTitle"
            style={{ paddingRight: '20px' }}
          >
            Field: {props.field}
          </span>
          <span
            className="content-panel-subTitle"
            style={{ paddingRight: '20px' }}
          >
            Aggregation method: {props.aggregationMethod}
          </span>
          <span className="content-panel-subTitle">
            State: {props.feature.featureEnabled ? 'Enabled' : 'Disabled'}
          </span>
        </Fragment>
      ) : (
        <Fragment>
          <span
            className="content-panel-subTitle"
            style={{ paddingRight: '20px' }}
          >
            Custom expression:{' '}
            <EuiLink onClick={() => setShowCustomExpression(true)}>
              View code
            </EuiLink>
          </span>
          <span className="content-panel-subTitle">
            State: {props.feature.featureEnabled ? 'Enabled' : 'Disabled'}
          </span>
        </Fragment>
      )}
    </EuiText>
  );

  const featureData = prepareDataForChart(props.featureData, props.dateRange);
  return (
    <ContentPanel
      title={
        props.feature.featureEnabled
          ? props.feature.featureName
          : `${props.feature.featureName} (disabled)`
      }
      bodyStyles={
        !props.feature.featureEnabled
          ? { backgroundColor: getDisabledChartBackground() }
          : {}
      }
      subTitle={featureDescription()}
      actions={
        props.edit ? <EuiButton onClick={props.onEdit}>Edit</EuiButton> : null
      }
    >
      {props.featureData.length > 0 ? (
        <div
          style={{
            height: '200px',
            width: '100%',
            opacity: showLoader ? 0.2 : 1,
          }}
        >
          <Chart>
            <Settings
              showLegend
              legendPosition={Position.Right}
              theme={[FEATURE_CHART_THEME]}
            />
            {props.feature.featureEnabled ? (
              <RectAnnotation
                dataValues={props.annotations || []}
                id="annotations"
                style={{
                  stroke: darkModeEnabled() ? 'red' : '#D5DBDB',
                  strokeWidth: 1,
                  opacity: 0.8,
                  fill: darkModeEnabled() ? 'red' : '#D5DBDB',
                }}
              />
            ) : null}
            <Axis
              id="left"
              title={props.featureDataSeriesName}
              position="left"
              showGridLines
            />
            <Axis id="bottom" position="bottom" tickFormat={timeFormatter} />
            <LineSeries
              id="featureData"
              name={props.featureDataSeriesName}
              xScaleType={ScaleType.Time}
              yScaleType={ScaleType.Linear}
              xAccessor={CHART_FIELDS.PLOT_TIME}
              yAccessors={[CHART_FIELDS.DATA]}
              color={[CHART_COLORS.FEATURE_DATA_COLOR]}
              data={featureData}
            />
          </Chart>
          {showCustomExpression ? (
            <CodeModal
              title={props.feature.featureName}
              subtitle="Custom expression"
              code={JSON.stringify(props.feature.aggregationQuery, null, 4)}
              getModalVisibilityChange={() => true}
              closeModal={() => setShowCustomExpression(false)}
            />
          ) : null}
        </div>
      ) : (
        <EuiEmptyPrompt
          style={{ maxWidth: '45em' }}
          body={
            <EuiText>
              <p>{`There is no data to display for feature ${props.feature.featureName}`}</p>
            </EuiText>
          }
        />
      )}
    </ContentPanel>
  );
};
