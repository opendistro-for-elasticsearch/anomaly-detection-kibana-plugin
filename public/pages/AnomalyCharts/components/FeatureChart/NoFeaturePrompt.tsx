/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';

import {
  EuiEmptyPrompt,
  EuiText,
  EuiFlexItem,
  EuiFlexGroup,
  EuiButton,
} from '@elastic/eui';

import ContentPanel from '../../../../components/ContentPanel/ContentPanel';

interface NoFeaturePromptProps {
  detectorId: string;
  title?: string;
}

export const NoFeaturePrompt = (props: NoFeaturePromptProps) => {
  return (
    <React.Fragment>
      <ContentPanel
        title={props.title ? props.title : ''}
        titleSize="xs"
        titleClassName="preview-title"
      >
        <EuiFlexGroup>
          <EuiFlexItem>
            <EuiEmptyPrompt
              body={
                <EuiText>
                  No features have been added to this anomaly detector. A
                  feature is a metric that used for anomaly detection. A
                  detector can discover anomalies across one or many features.
                  This system reports an anomaly score based on how strong a
                  signal might be.
                </EuiText>
              }
              actions={[
                <EuiButton
                  data-test-subj="createButton"
                  href={`#/detectors/${props.detectorId}/features`}
                >
                  Add feature
                </EuiButton>,
              ]}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </ContentPanel>
    </React.Fragment>
  );
};
