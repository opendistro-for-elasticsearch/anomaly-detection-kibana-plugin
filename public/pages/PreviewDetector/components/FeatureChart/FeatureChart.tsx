/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  Chart,
  Axis,
  LineSeries,
  RectAnnotation,
  niceTimeFormatter,
  Settings,
} from '@elastic/charts';
import { EuiButton, EuiEmptyPrompt, EuiText } from '@elastic/eui';
import React from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { useDelayedLoader } from '../../../../hooks/useDelayedLoader';
import { Moment } from 'moment';
import { FeatureAggregationData } from 'public/models/interfaces';
import { darkModeEnabled } from '../../../../utils/kibanaUtils';

interface FeatureChartProps {
  title: string;
  enabled: boolean;
  onEdit(ev: React.MouseEvent<HTMLButtonElement>): void;
  isEdit: boolean;
  featureData: FeatureAggregationData[];
  annotations: any[];
  isLoading: boolean;
  startDateTime: Moment;
  endDateTime: Moment;
}
const getDisabledChartBackground = () =>
  darkModeEnabled() ? '#25262E' : '#F0F0F0';

export const FeatureChart = (props: FeatureChartProps) => {
  const timeFormatter = niceTimeFormatter([
    props.startDateTime.valueOf(),
    props.endDateTime.valueOf(),
  ]);
  const showLoader = useDelayedLoader(props.isLoading);

  return (
    <ContentPanel
      title={props.enabled ? props.title : `${props.title} ( disabled )`}
      titleSize="xs"
      titleClassName="preview-title"
      panelStyles={props.isEdit ? { border: '5px solid #96C8DA' } : {}}
      bodyStyles={
        !props.enabled ? { backgroundColor: getDisabledChartBackground() } : {}
      }
      actions={
        <EuiButton
          onClick={props.onEdit}
          disabled={props.isEdit}
          data-test-subj="editFeature"
        >
          Edit
        </EuiButton>
      }
    >
      {props.featureData.length > 0 ? (
        <div
          style={{
            height: '300px',
            width: '100%',
            opacity: showLoader ? 0.2 : 1,
          }}
        >
          <Chart>
            <Settings
              showLegend
              legendPosition="bottom"
            />
            {props.enabled ? (
              <RectAnnotation
                dataValues={props.annotations || []}
                id="annotations"
                // annotationId={getAnnotationId('react')}
                style={{
                  stroke: darkModeEnabled() ? 'red' : '#FCAAAA',
                  strokeWidth: 1,
                  opacity: 0.8,
                  fill: darkModeEnabled() ? 'red' : '#FCAAAA',
                }}
              />
            ) : null}
            <Axis id="left" title={props.title} position="left" />
            <Axis id="bottom" position="bottom" tickFormat={timeFormatter} />
            <LineSeries
              id="lines"
              name={`Aggregated data for ${props.title}`}
              xScaleType="time"
              yScaleType="linear"
              xAccessor={'startTime'}
              yAccessors={['data']}
              data={props.featureData}
            />
          </Chart>
        </div>
      ) : (
        <EuiEmptyPrompt
          style={{ maxWidth: '45em' }}
          body={
            <EuiText>
              <p>{`There is no data to display for feature ${props.title}`}</p>
            </EuiText>
          }
        />
      )}
    </ContentPanel>
  );
};
