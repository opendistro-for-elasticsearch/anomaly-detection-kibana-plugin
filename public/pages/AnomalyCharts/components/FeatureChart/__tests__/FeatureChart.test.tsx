/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { FeatureChart } from '../FeatureChart';
import moment from 'moment';
import { FEATURE_TYPE } from '../../../../../models/interfaces';

const feature = {
  featureId: 'feature-id',
  featureName: 'feature-name',
  featureEnabled: true,
  importance: 1,
  aggregationQuery: {
    F1: {
      avg: {
        field: 'value',
      },
    },
  },
};

const featureData = [
  { data: 0.8, endTime: 1587453273000, startTime: 1587446073000 },
];
const dateRange = {
  startDate: moment(1587431968000),
  endDate: moment(1587453568000),
};

describe('<FeatureChart /> spec', () => {
  describe('Feature Chart', () => {
    test('renders component with simple aggregation', () => {
      const { container } = render(
        <FeatureChart
          feature={feature}
          featureData={featureData}
          annotations={[]}
          isLoading={false}
          dateRange={dateRange}
          featureType={FEATURE_TYPE.SIMPLE}
          field="value"
          aggregationMethod="sum"
          featureDataSeriesName="anomaly grade"
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders component with custom expression', () => {
      const { container } = render(
        <FeatureChart
          feature={feature}
          featureData={featureData}
          annotations={[]}
          isLoading={false}
          dateRange={dateRange}
          featureType={FEATURE_TYPE.CUSTOM}
          aggregationQuery={JSON.stringify(feature.aggregationQuery)}
          featureDataSeriesName="anomaly grade"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
