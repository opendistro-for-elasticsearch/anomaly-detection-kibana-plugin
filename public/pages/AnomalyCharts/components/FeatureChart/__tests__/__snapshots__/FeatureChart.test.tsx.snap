// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeatureChart /> spec Feature Chart renders component with custom expression 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small content-panel-title"
        >
          feature-name
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--small"
            >
              <p
                style="font-size: 12px;"
              >
                Custom expression:
                 
                <button
                  class="euiLink euiLink--primary"
                  type="button"
                >
                  View code
                </button>
                ; State: 
                Enabled
                 
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        style="height: 200px; width: 100%; opacity: 1;"
      >
        <div
          class="echChart"
        >
          <div
            class="echChartStatus"
            data-ech-render-complete="true"
            data-ech-render-count="1"
          />
          <div
            class="echChartResizer"
          />
          <div
            class="echLegend echLegend--right"
          >
            <div
              class="echLegendListContainer"
              style="width: 30px; max-width: 30px;"
            >
              <div
                class="echLegendList"
                style="padding-top: 10px; padding-bottom: 10px;"
              >
                <div
                  class="echLegendItem echLegendItem--right"
                >
                  <div
                    aria-label="series color"
                    class="echLegendItem__color"
                    title="series color"
                  >
                    <svg
                      class="echIcon"
                      color="#16191F"
                      focusable="false"
                      height="16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <defs>
                        <circle
                          cx="8"
                          cy="8"
                          id="dot-a"
                          r="4"
                        />
                      </defs>
                      <g>
                        <use
                          xlink:href="#dot-a"
                        />
                      </g>
                    </svg>
                  </div>
                  <div
                    class="echLegendItem__label"
                    title="anomaly grade"
                  >
                    anomaly grade
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="echContainer"
          >
            <div
              class="echChartPointerContainer"
              style="cursor: default;"
            >
              <div
                class="echCrosshair"
              />
              <svg
                class="echHighlighter"
              >
                <defs>
                  <clippath
                    id="echHighlighterClipPath__9627cc98-cf70-44c5-8bbd-8b96bddb2798"
                  >
                    <rect
                      height="0"
                      width="0"
                      x="0"
                      y="0"
                    />
                  </clippath>
                </defs>
                <g
                  transform="translate(0, 0) rotate(0)"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FeatureChart /> spec Feature Chart renders component with simple aggregation 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small content-panel-title"
        >
          feature-name
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--small"
            >
              <p
                style="font-size: 12px;"
              >
                Field: 
                value
                ; Aggregation method: 
                sum
                ; State: 
                Enabled
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        style="height: 200px; width: 100%; opacity: 1;"
      >
        <div
          class="echChart"
        >
          <div
            class="echChartStatus"
            data-ech-render-complete="true"
            data-ech-render-count="1"
          />
          <div
            class="echChartResizer"
          />
          <div
            class="echLegend echLegend--right"
          >
            <div
              class="echLegendListContainer"
              style="width: 30px; max-width: 30px;"
            >
              <div
                class="echLegendList"
                style="padding-top: 10px; padding-bottom: 10px;"
              >
                <div
                  class="echLegendItem echLegendItem--right"
                >
                  <div
                    aria-label="series color"
                    class="echLegendItem__color"
                    title="series color"
                  >
                    <svg
                      class="echIcon"
                      color="#16191F"
                      focusable="false"
                      height="16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <defs>
                        <circle
                          cx="8"
                          cy="8"
                          id="dot-a"
                          r="4"
                        />
                      </defs>
                      <g>
                        <use
                          xlink:href="#dot-a"
                        />
                      </g>
                    </svg>
                  </div>
                  <div
                    class="echLegendItem__label"
                    title="anomaly grade"
                  >
                    anomaly grade
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="echContainer"
          >
            <div
              class="echChartPointerContainer"
              style="cursor: default;"
            >
              <div
                class="echCrosshair"
              />
              <svg
                class="echHighlighter"
              >
                <defs>
                  <clippath
                    id="echHighlighterClipPath__d2536731-6d91-4094-a106-023a8f2bf6e1"
                  >
                    <rect
                      height="0"
                      width="0"
                      x="0"
                      y="0"
                    />
                  </clippath>
                </defs>
                <g
                  transform="translate(0, 0) rotate(0)"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
